spi/imx6ull-qspi/qspi.c

```diff
@@ -0,0 +1,326 @@
+	{ { pctl_pad_nand_d7, 1, 4, 0 }, { pctl_pad_nand_ale, 1, 4, 0 }, { pctl_pad_nand_wp, 1, 4, 0 }, { pctl_pad_nand_rdy, 1, 4, 0 }, { pctl_pad_nand_ce0, 1, 4, 0 }, { pctl_pad_nand_ce1, 1, 4, 0 }, { pctl_pad_nand_cle, 1, 4, 0 }, { pctl_pad_nand_dqs, 1, 4, 0 } },
+	{ { pctl_pad_nand_re, 1, 4, 0 }, { pctl_pad_nand_we, 1, 4, 0 }, { pctl_pad_nand_d0, 1, 4, 0 }, { pctl_pad_nand_d1, 1, 4, 0 }, { pctl_pad_nand_d2, 1, 4, 0 }, { pctl_pad_nand_d3, 1, 4, 0 }, { pctl_pad_nand_d4, 1, 4, 0 }, { pctl_pad_nand_d5, 1, 4, 0 } },
+};
+
+
+struct {
+	volatile uint32_t *base;
+	uint32_t flash_base_addr[2];
+	bool init;
+} qspi_common = { .base = NULL, .init = false, .flash_base_addr = { QSPI_MMAP_BASE, QSPI_MMAP_BASE + QSPI_FLASH_A1_SIZE + QSPI_FLASH_A2_SIZE } };
+
+// TODO ustawić ile ohmów w rejestrze na flashu.
+// TODO find clock frequency - ~50Mhz
+
+static void set_mux(qspi_dev_t dev_no)
+{
+	platformctl_t ctl;
+	int i;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_iomux;
+
+	for (i = 0; i < 8; i++) {
+		ctl.iomux.mux = qspi_pctl_mux[dev_no][i].mux;
+		ctl.iomux.mode = qspi_pctl_mux[dev_no][i].val;
+		ctl.iomux.sion = qspi_pctl_mux[dev_no][i].sion;
+		platformctl(&ctl);
+	}
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_iopad;
+
+	for (i = 0; i < 8; i++) {
+		ctl.iopad.pad = qspi_pctl_pad[dev_no][i].pad;
+		ctl.iopad.speed = qspi_pctl_pad[dev_no][i].speed;
+		ctl.iopad.dse = qspi_pctl_pad[dev_no][i].dse;
+		ctl.iopad.sre = qspi_pctl_pad[dev_no][i].sre;
+		platformctl(&ctl);
+	}
+}
+
+static void set_clk()
+{
+	platformctl_t ctl;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_devclock;
+	ctl.devclock.dev = pctl_clk_qspi;
+	ctl.devclock.state = 0x03;
+
+	platformctl(&ctl);
+}
+
+int qspi_setLutSeq(const lut_seq_t *lut, unsigned int lut_seq)
+{
+	if (lut_seq >= NUM_LUT_SEQ) {
+		return -1;
+	}
+	if (*QSPI_LCKCR_ADDR & QSPI_LCKCR_LOCK) {
+		*QSPI_LUT_KEY_ADDR = QSPI_LUT_KEY;
+		*QSPI_LCKCR_ADDR = (*QSPI_LCKCR_ADDR & ~(QSPI_LCKCR_UNLOCK | QSPI_LCKCR_LOCK)) | QSPI_LCKCR_UNLOCK;
+	}
+
+	for (int i = 0; i < 4; i++) {
+		*QSPI_LUTn_ADDR(QSPI_SEQ_START_REGN(lut_seq) + i) = (((uint32_t)lut->instrs[2 * i + 1]) << 16) | lut->instrs[2 * i];
+	}
+
+	*QSPI_LUT_KEY_ADDR = QSPI_LUT_KEY;
+	*QSPI_LCKCR_ADDR = (*QSPI_LCKCR_ADDR & ~(QSPI_LCKCR_UNLOCK | QSPI_LCKCR_LOCK)) | QSPI_LCKCR_LOCK;
+}
+
+static void set_IPCR(int seq_num, size_t idatsz)
+{
+	*QSPI_IPCR_ADDR = QSPI_IPCR_SET_IDATSZ(QSPI_IPCR_SET_SEQID(*QSPI_IPCR_ADDR, seq_num), idatsz & 0xFFFF);
+}
+
+void qspi_setTCSH(uint8_t cycles)
+{
+	*QSPI_FLSHCR_ADDR = (*QSPI_FLSHCR_ADDR & ~(0x0F << 8)) | (cycles << 8);
+}
+
+void qspi_setTCSS(uint8_t cycles)
+{
+	*QSPI_FLSHCR_ADDR = (*QSPI_FLSHCR_ADDR & ~(0x0F)) | cycles;
+}
+
+int _qspi_readBusy(qspi_dev_t dev, unsigned int lut_seq, uint32_t addr, void *buf, size_t size)
+{
+	uint8_t to_read, byte, i;
+	uint16_t len = 0;
+	uint32_t reg;
+	int err;
+
+
+	if (dev != qspi_flash_a && dev != qspi_flash_b) {
+		return -ENODEV;
+	}
+	if (lut_seq >= NUM_LUT_SEQ || size > MAX_READ_LEN) {
+		return -1;
```

Comment:
Either use generic -1 for error or errno code, but be consistent 
```suggestion
		return -EINVAL;
```